src/librustc_mir/build/scope.rs

```diff
@@ -455,25 +458,24 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
         };
 
         for scope in self.scopes.iter_mut().rev() {
-            if scope.extent == extent {
+            // We must invalidate all the caches leading up to the scope we’re looking for, because
+            // the cached blocks will branch into build of scope not containing the new drop. If we
+            // add stuff to the currently inspected scope, then in some cases the non-unwind caches
+            // may become invalid, therefore we should invalidate these as well. The unwind caches
+            // will stay correct, because the already generated unwind blocks cannot be influenced
+            // by just added drop.
+            let this_scope = scope.extent == extent;
+            scope.invalidate_cache(!this_scope);
```

Comment:
This will invalidate the unwind cache in cases other than `DropKind::Value`, which is undesirable (as `StorageDead` doesn't end up on the unwind path at all).